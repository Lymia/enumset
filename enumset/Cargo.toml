[package]
name = "enumset"
version = "1.1.5"
authors = ["Alissa Rao <lymia@lymiahugs.com>"]
edition = "2021"
rust-version = "1.61"

description = "A library for creating compact sets of enums."
keywords = ["enum", "bitset"]
categories = ["data-structures"]

documentation = "https://docs.rs/enumset/"
repository = "https://github.com/Lymia/enumset"
readme = "README.md"

license = "MIT/Apache-2.0"

[badges]
travis-ci = { repository = "Lymia/enumset", branch = "master" }
maintenance = { status = "actively-developed" }

[features]
serde = ["serde2", "enumset_derive/serde"]
alloc = []
proc-macro-crate = ["enumset_derive/proc-macro-crate"]

# Deprecated features
std = ["alloc", "enumset_derive/proc-macro-crate", "enumset_derive/std_deprecation_warning"]

[package.metadata.docs.rs]
all-features = true
rustdoc-args = ["--cfg", "docsrs"]

[dependencies]
enumset_derive = { version = "0.10.0", path = "../enumset_derive" }
serde2 = { package = "serde", version = "1", default-features = false, optional = true }

[dev-dependencies]
bincode = { version = "1", features = ["i128"] }
serde_derive = "1"
serde_json = "1"
