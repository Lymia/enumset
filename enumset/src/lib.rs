#![no_std]
#![deny(missing_docs)]
#![allow(clippy::missing_safety_doc)] // The safety requirement is "use the procedural derive".
#![allow(clippy::needless_range_loop)] // range loop style is clearer in most places in enumset
#![cfg_attr(docsrs, feature(doc_cfg))]

//! A library for defining enums that can be used in compact bit sets. It supports arbitrarily
//! large enums, and has very basic support for using them in constants.
//!
//! # Cargo Features
//!
//! The following cargo features are available for this crate:
//!
//! * `serde`: Allows serialization and deserialization of the types in this crate.
//! * `alloc`: Enables the use of functions that requires an allocator.
//! * `proc-macro-crate`: Enable the use of the `proc-macro-crate` crate to allow the renaming of
//!   the `enumset` crate in your user crate. This feature increases the MSRV to 1.69.0
//!
//! # Defining enums for use with EnumSet
//!
//! Enums to be used with [`EnumSet`] should be defined using `#[derive(EnumSetType)]`:
//!
//! ```rust
//! # use enumset::*;
//! #[derive(EnumSetType, Debug)]
//! pub enum Enum {
//!    A, B, C, D, E, F, G,
//! }
//! ```
//!
//! For more information on more advanced use cases, see the documentation for
//! [`#[derive(EnumSetType)]`](./derive.EnumSetType.html).
//!
//! # Working with EnumSets
//!
//! EnumSets can be constructed via [`EnumSet::new()`] like a normal set. In addition,
//! `#[derive(EnumSetType)]` creates operator overloads that allow you to create EnumSets like so:
//!
//! ```rust
//! # use enumset::*;
//! # #[derive(EnumSetType, Debug)] pub enum Enum { A, B, C, D, E, F, G }
//! let new_set = Enum::A | Enum::C | Enum::G;
//! assert_eq!(new_set.len(), 3);
//! ```
//!
//! All bitwise operations you would expect to work on bitsets also work on both EnumSets and
//! enums with `#[derive(EnumSetType)]`:
//! ```rust
//! # use enumset::*;
//! # #[derive(EnumSetType, Debug)] pub enum Enum { A, B, C, D, E, F, G }
//! // Intersection of sets
//! assert_eq!((Enum::A | Enum::B) & Enum::C, EnumSet::empty());
//! assert_eq!((Enum::A | Enum::B) & Enum::A, Enum::A);
//! assert_eq!(Enum::A & Enum::B, EnumSet::empty());
//!
//! // Symmetric difference of sets
//! assert_eq!((Enum::A | Enum::B) ^ (Enum::B | Enum::C), Enum::A | Enum::C);
//! assert_eq!(Enum::A ^ Enum::C, Enum::A | Enum::C);
//!
//! // Difference of sets
//! assert_eq!((Enum::A | Enum::B | Enum::C) - Enum::B, Enum::A | Enum::C);
//!
//! // Complement of sets
//! assert_eq!(!(Enum::E | Enum::G), Enum::A | Enum::B | Enum::C | Enum::D | Enum::F);
//! ```
//!
//! The [`enum_set!`] macro allows you to create EnumSets in constant contexts:
//!
//! ```rust
//! # use enumset::*;
//! # #[derive(EnumSetType, Debug)] pub enum Enum { A, B, C, D, E, F, G }
//! const CONST_SET: EnumSet<Enum> = enum_set!(Enum::A | Enum::B);
//! assert_eq!(CONST_SET, Enum::A | Enum::B);
//! ```
//!
//! Mutable operations on the [`EnumSet`] otherwise similarly to Rust's builtin sets:
//!
//! ```rust
//! # use enumset::*;
//! # #[derive(EnumSetType, Debug)] pub enum Enum { A, B, C, D, E, F, G }
//! let mut set = EnumSet::new();
//! set.insert(Enum::A);
//! set.insert_all(Enum::E | Enum::G);
//! assert!(set.contains(Enum::A));
//! assert!(!set.contains(Enum::B));
//! assert_eq!(set, Enum::A | Enum::E | Enum::G);
//! ```

#[cfg(feature = "alloc")]
extern crate alloc;

mod macros;

#[macro_use]
mod set_common;

mod derive;
mod record;
mod repr;
mod set;
mod set_mixed;
mod traits;

pub use crate::derive::EnumSetType;
pub use crate::macros::__internal;
pub use crate::record::{EnumRecord, EnumRecordIntoIter, EnumRecordIter, EnumRecordIterMut};
pub use crate::set::{EnumSet, EnumSetIter};
pub use crate::set_mixed::{MixedEnumSet, MixedEnumSetIter, MixedValue};
pub use crate::traits::{EnumSetType, EnumSetTypeWithRepr};
