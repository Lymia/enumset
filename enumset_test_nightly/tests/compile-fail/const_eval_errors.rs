use enumset::*;

#[derive(EnumSetType)]
enum BadVariants {
    Variant = -(const { 1 + 1 }),
}

#[derive(EnumSetType)]
enum BadVariants2 {
    Variant = !(1 > 2),
}

#[derive(EnumSetType)]
#[repr(u64)]
enum BadVariants3 {
    Variant = (1u64 << 0x100000000u64) + 1,
}

#[derive(EnumSetType)]
#[repr(u64)]
enum BadVariants4 {
    Variant = 1 + (1u64 >> 0x100000000u64),
}

#[derive(EnumSetType)]
enum BadVariants5 {
    Variant = "test",
}

#[derive(EnumSetType)]
enum BadVariants6 {
    Variant = *(&5),
}

#[derive(EnumSetType)]
#[repr(u128)]
enum BadVariants7 {
    Variant = 0x10000000000000000u128,
}

fn main() {}