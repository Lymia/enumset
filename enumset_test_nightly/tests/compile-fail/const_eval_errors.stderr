error: Expression not supported by enumset.
 --> tests/compile-fail/const_eval_errors.rs:5:17
  |
5 |     Variant = -(const { 1 + 1 }),
  |                 ^^^^^^^^^^^^^^^

error: Expression not supported by enumset.
  --> tests/compile-fail/const_eval_errors.rs:10:17
   |
10 |     Variant = !(1 > 2),
   |                 ^^^^^

error: Error while evaluating discriminator.
  --> tests/compile-fail/const_eval_errors.rs:16:16
   |
16 |     Variant = (1u64 << 0x100000000u64) + 1,
   |                ^^^^^^^^^^^^^^^^^^^^^^

error: Error while evaluating discriminator.
  --> tests/compile-fail/const_eval_errors.rs:22:20
   |
22 |     Variant = 1 + (1u64 >> 0x100000000u64),
   |                    ^^^^^^^^^^^^^^^^^^^^^^

error: Expression not supported by enumset.
  --> tests/compile-fail/const_eval_errors.rs:27:15
   |
27 |     Variant = "test",
   |               ^^^^^^

error: Expression not supported by enumset.
  --> tests/compile-fail/const_eval_errors.rs:32:15
   |
32 |     Variant = *(&5),
   |               ^^^^^

error: Enum discriminants must fit into `isize`.
  --> tests/compile-fail/const_eval_errors.rs:38:15
   |
38 |     Variant = 0x10000000000000000u128,
   |               ^^^^^^^^^^^^^^^^^^^^^^^

error[E0308]: mismatched types
  --> tests/compile-fail/const_eval_errors.rs:10:15
   |
10 |     Variant = !(1 > 2),
   |               ^^^^^^^^ expected `isize`, found `bool`

error[E0080]: attempt to shift left by `4294967296_u64`, which would overflow
  --> tests/compile-fail/const_eval_errors.rs:16:15
   |
16 |     Variant = (1u64 << 0x100000000u64) + 1,
   |               ^^^^^^^^^^^^^^^^^^^^^^^^ evaluation of `BadVariants3::Variant::{constant#0}` failed here

error[E0080]: attempt to shift right by `4294967296_u64`, which would overflow
  --> tests/compile-fail/const_eval_errors.rs:22:19
   |
22 |     Variant = 1 + (1u64 >> 0x100000000u64),
   |                   ^^^^^^^^^^^^^^^^^^^^^^^^ evaluation of `BadVariants4::Variant::{constant#0}` failed here

error[E0308]: mismatched types
  --> tests/compile-fail/const_eval_errors.rs:27:15
   |
27 |     Variant = "test",
   |               ^^^^^^ expected `isize`, found `&str`
